gridded_etl_tools/utils/extractor.py

```diff
@@ -0,0 +1,363 @@
+                    f" , attempt {counter}"
+                )
+                counter += 1
+        else:
+            self.dm.info(f"Couldn't find or download a remote file for {informative_id}")
+            raise FileNotFoundError(f"Too many ({counter}) failed download attempts from server")
+
+
+class FTPExtractor(Extractor):
+    """
+    Create an object that provides an interface to a climate data source's FTP server. The object is able to open its
+    connection within a context manager, navigate to specific working directory, match files located in subdirectories,
+    and fetch files to a given destination folder.
+
+    Currently only anonymous FTP access is supported.
+    """
+
+    ftp: ftplib.FTP
+    host: str
+
+    # Used to limit retrieval to a single thread at a time
+    semaphore: multiprocessing.synchronize.Semaphore = multiprocessing.Semaphore()
+
+    def __enter__(self) -> FTPExtractor:
+        """
+        Open a connection to the FTP server at this object's given source from within a context manager. When creating
+        a context, pass the host address unless `FTPExtractor.host` was already set.
+
+        Example
+        -------
+        my_extractor = FTPExtractor(my_dataset_manager)
+        with my_extractor("ftp.cdc.noaa.gov") as extractor:
+            # get source files
+
+        Returns
+        -------
+        FTPConnection
+            this object
+
+        Raises
+        ------
+        ValueError
+            If FTPExtractor.
+        """
+        if not hasattr(self, "host"):
+            raise ValueError("FTPExtractor must have a host parameter to open connection")
+        else:
+            self.dm.info(f"Opening a connection to {self.host}")
+            self.ftp = ftplib.FTP(self.host)
+            self.ftp.login()
+            return self
+
+    def __exit__(self, *exception):
+        """
+        Close the connection with the FTP server at this object's given source. This will be called automatically when
+        exiting the connection context manager.
+
+        Parameters
+        ----------
+        *exception
+            Exception information passed automatically by Python
+        """
+        self.ftp.close()
+        self.dm.info(f"Closed connection to {self.host}")
+
+    def __call__(self, host: str):
+        """
+        When called like a function, set the host parameter. See `FTPExtractor.__enter__`.
+
+        Parameters
+        ----------
+        host
+            Address to connect to for source data
+        """
+        self.host = host
+        return self
+
+    @property
+    def cwd(self) -> pathlib.Path:
+        """
+        Returns
+        -------
+        pathlib.Path
+            The object's working directory on the FTP server
+
+        Raises
+        ------
+        RuntimeError
+            If the FTP connection is not open yet
+        """
+        try:
+            return pathlib.Path(self.ftp.pwd())
```

Comment:
I don't think we should use `pathlib.Path`s to represent the paths on the FTP server. If someone were to run this code on windows, `str(path)` would concatenate the parts of the path with backslashes, resulting in an invalid FTP path. I would use `pathlib.PurePosixPath`